### ISSUE TITLE
Problem reading PDS4 file with skipped byte in record line

### ISSUE THREAD

**Issue by kalosma** [Small-Bodies-Node/PDS4-IDL#11]
We are using the PDS4-IDL code for reading some PDS4 files we are creating.
We have been able to read those products in the past, but then we had to make a change to a DateTime field because it had a bad format, it had a space in the middle between date and time (e.g. 2011-04-10 00:31:05.694), so we splitted it into two fields, one for the DATE and one for the TIME. The date field "ends" before the space while the time field starts after the space (e.g. time field finishes @byte 65574 while date field starts @byte 65576).

```
[...]
<Field_Binary>
                    <name>UTC_MIDPOINT_MET_DATE</name>
                    <field_number>11</field_number>
                    <field_location unit="byte">65565</field_location>
                    <data_type>ASCII_Date_YMD</data_type>
                    <field_length unit="byte">10</field_length>                
                    <description>MIDPOINT_MET converted to UTC Date, stored as
                        yyyy-mm-dd</description>
</Field_Binary>
<Field_Binary>
                    <name>UTC_MIDPOINT_MET_TIME</name>
                    <field_number>12</field_number>
                    <field_location unit="byte">65576</field_location>
                    <data_type>ASCII_Time</data_type>
                    <field_length unit="byte">12</field_length>                
                    <description>MIDPOINT_MET converted to UTC Time, stored as
                        hh:mm:ss.sss</description>
</Field_Binary>
[...]
```

There is a byte in the middle between Date and Time that is skippied because not needed. That is causing issues with the pds_read IDL code. 

Please can you fix this issue?

Here is a sample product.
[sample.zip](https://github.com/Small-Bodies-Node/PDS4-IDL/files/2274158/sample.zip)

**Comment by mkelley** [comment]
Thanks for the feedback and your patience while I was on vacation.

@eshaya Do you have an idea how to address this?

**Comment by kalosma** [comment]
I have not digged into the read_pds code, but it seems like the 'field_location' information for the binary field is not used and the start position to read for a given field is computed adding the previous field length to the previous field start position

**Comment by LevN0** [comment]
That sounds like a likely guess. Most likely the data read in IDL would be done by creating a structure and then reading the file all at once. This could most easily be addressed, I think, by looking at discontinuities in field_location and adding fake fields for the read, then deleting them afterward.

**Comment by mkelley** [comment]
Thanks, @LevN0 .

I checked with our standards expert, having dead bytes in the file is allowed but not recommended.  But because it is allowed, we should still address the issue.

**Comment by eshaya** [comment]
Yes,  I plan on looking at this and fixing it.  I have been putting out 
several other fires lately, but I may get to it today.

Ed

**Comment by eshaya** [comment]
I have looked at this code issue and have decided not to make changes.  
I guess here is as good a place as any to describe the complex trade off 
issues that are involved.  Therefore, I will make this a longish 
detailed explanation.  But, if you are not interested in the nitty 
gritty,  you just need to know that, for now, the read_pds ignores 
locations in binary data.   As Lev correctly pointed out, it uses 
unformatted read (readu) for binary data which has no format statement 
or template, and expects to simply read in the structure with a 
continuous read.  If there are spaces between fields, the user will need 
to add dummy variables to fill them.

Here are the various options that were faced when this was originally 
written:

1) Use readu:  This is fastest and allows for any complicated structure, 
including nested structures. Although the code for nested structure 
(tables embedded with other fields) is not actually there. The plan, 
before funding was lost, was to add that capability.  Relatively simple 
code and easy to maintain.  The only real problem is this case, where 
there are blank bytes between fields, but since most binary data in the 
world is written with unformatted write, it would be rare to see blank 
bytes.  And there is a simple fix on the user side.

2) Use readu and programmatically add dummy fields for blank bytes:   
Adding fields is not hard.  There is some overhead in assuring that they 
have unique names, but it may be best not to surprise the user with 
unexpected fields of mysterious meaning. To fix that, one could remove 
the blank byte fields before return.  However, removing tags in 
structures is not allowed in IDL.  So, one has to build a new structure 
and then transfer all data from the old structure.  That doubles both 
cpu time and memory usage of the program.  Usually, this is not so bad, 
but for very large data sets it could be a limiting factor. 
Interestingly though, if one used Lists instead of Structures, then one 
can quickly and easily remove fields.  So, I would recommend going that 
route.  However, this is a major change and would not be backwards 
compatible with IDL below version 8.2.

3) Use READ_BINARY: This routine reads binary data and expects a 
binary_template which specifies the datatypes and locations of the 
fields.   It is not more or less complicated as readu because filling 
out the template is about as hard as creating the structure. It works by 
moving the pointer to the location of each field and reading one field 
at a time.  It is slower, but the main difficulty is that the template 
only describes a structure with scalars and arrays.  So, ultimately it 
is a dead end for PDS because it cannot describe the full set of 
possibilities.

4) Write our own READ_BINARY:  We could write code to add whatever 
complexity is needed to the templates (allow templates within templates 
etc.)  and then rewrite read_binary to be more general.  We did not go 
this route, both because it is slow and it adds many more lines of code 
(read, it would cost more to write and maintain).  If you are seeking 
version 7 compatibility and full PDS4, I would think deleting blank 
fields from a readu read is easier.

5)  Read the file into a single array of bytes:  Here you read the file 
into an array in memory and then manipulate the array and then transform 
it into the structure.  This would allow for any general structure 
possible in IDL, but it would require a lot of code.  Basically, you are 
throwing out all of the effort of the IDL code writers who have 
optimized code for reading files, so my guess is that this would end up 
being quite slow.  It surely would be the most difficult to write.

In conclusion, I see the best option is to keep things as they are, and 
when there is funding for read_pds for version 8, make use of Lists so 
that one can continue to use readu, and to make use of the new parse_xml.

Ed

**Comment by mkelley** [comment]
@kalosma If you are allowed to change the data, a fix could be to replace the space with a 'T' and use a single field with the format ASCII_Date_Time_YMD_UTC: YYYY-MM-DDTHH:MM:SS.SSSZ.

**Comment by LevN0** [comment]
>However, removing tags in structures is not allowed in IDL.  So, one has to build a new structure and then transfer all data from the old structure.  That doubles both cpu time and memory usage of the program.  Usually, this is not so bad, but for very large data sets it could be a limiting factor.

I would not think that transferring memory would double, or seriously increase, CPU time, as compared to reading from disk. Quick, albeit not entirely thorough, tests show that transferring data between structures is quite quick with-in reasonable memory limits (e.g. 32 GB). However, memory indeed seems like a catch - at least off the top of my head I do not see an easy solution with structures. That said, I would think the trade off is worth it because: 1. as you pointed out, tables where the need to use this solution and thus double the RAM are quite rare (and all other tables can skip it) and, 2. tables (but not arrays/images) where doubling them in size would cause RAM issues are vanishingly rare, especially in PDS. Thus it takes two very rare situations combined to cause an issue, but allows standards compliant reading.

**Comment by kalosma** [comment]
> @kalosma If you are allowed to change the data, a fix could be to replace the space with a 'T' and use a single field with the format ASCII_Date_Time_YMD_UTC: YYYY-MM-DDTHH:MM:SS.SSSZ.

Of course that would be the best solution. Unfortunately the binary data comes from a previous PDS release (in PDS 3 format), we are adding a PDS 4 label and minimized the changes to the binary files to avoid introducing errors. Since PDS 4 format supports that, the simplest solution to fix the date time format issue in our new PDS 4 product was to split it into two fields just in the label, without modifying the binary file itself.

**Comment by eshaya** [comment]
Lev,

     You may or may not be right about the transfer time of a 
structure.  Are you talking about tests within IDL or a compiled 
language, because timing in an interpreted language is more complicated 
to assess?   Most of the time is in interpretation, especially bad are 
do loops.  The transfer would require at least one do loop through the 
tags and more if there were arrays. Double that, to both create the 
structure and to fill them.  In comparison, I am sure that the entire 
readu procedure is written in a compiled language and probably optimized 
in spots in assembly.    But, the read time may not be a big issue.  I 
simply do not know if there exist huge data files in PDS or not.

However, I decided this is not that big a problem after all.  I realized 
later that we could just make people aware that, if they have a very 
large dataset they should be careful to not have gaps, and then the step 
of transferring the data would be skipped.

  But now the problem is that this is not just a  bug fix, it is a 
substantial amount of new code.   It is not clear when I will have the 
spare time to do it or if it is really needed.

Ed

**Comment by eshaya** [comment]
Calogero,

     I do not know why it was suggested to change the data.  Just add 
another field to the xml.  I have already done and it checked out fine.  
I did not bother updating all of the field_numbers because the code does 
not use them, but you can.

          <Field_Binary>
                     <name>UTC_MIDPOINT_MET_DATE</name>
                     <field_number>11</field_number>
                     <field_location unit="byte">65565</field_location>
<data_type>ASCII_Date_YMD</data_type>
                     <field_length unit="byte">10</field_length>
                     <description>MIDPOINT_MET converted to UTC Date, 
stored as
                         yyyy-mm-dd</description>
                 </Field_Binary>
                 <Field_Binary>
                     <name>Blank_1</name>
                     <field_number>12</field_number>
                     <field_location unit="byte">65575</field_location>
                     <data_type>ASCII_String</data_type>
                     <field_length unit="byte">1</field_length>
                     <description>blank</description>
                 </Field_Binary>
                 <Field_Binary>
                     <name>UTC_MIDPOINT_MET_TIME</name>
                     <field_number>12</field_number>
                     <field_location unit="byte">65576</field_location>
                     <data_type>ASCII_Time</data_type>
                     <field_length unit="byte">12</field_length>
                     <description>MIDPOINT_MET converted to UTC Time, 
stored as
                         hh:mm:ss.sss</description>
                 </Field_Binary>

Ed

**Comment by kalosma** [comment]
Ed,

thanks for you suggestion we can do that

**Comment by mkelley** [comment]
I think this specific case is resolved, so I am closing the issue.  If the general case is still an issue, then we can reopen or create a new issue dedicated to it.
